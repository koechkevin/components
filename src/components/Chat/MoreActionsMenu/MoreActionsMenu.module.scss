@import '../../../styles/global';

@mixin item {
  &:hover {
    color: $color-white;
    background-color: $color-primary;
  }

  &:global(.ant-menu-item) {
    margin-bottom: -5px;
    font-size: 14px;

    &:last-child {
      margin-bottom: 4px;
    }
  }

  &:global(.ant-menu-item-selected) {
    color: $color-primary;
    background-color: $color-white;
  }
}

@mixin responsive-item {
  &:hover {
    background: $color-primary-light;
    color: $color-primary;
  }

  &:global(.ant-menu-item-selected) {
    color: $color-primary;
    background: $color-primary-light !important;
  }
}

.root {
  margin: 5px 0;
  border-radius: 5px;
  background-color: $color-white;
  border: 1px solid $color-gray-dark;

  .item {
    @include item;
  }

  .danger {
    @include item;

    &:global(.ant-menu-item-selected) {
      color: $color-white;
      background: $color-error !important;
    }

    &:global(.ant-menu-item) {
      &:hover {
        color: $color-white;
        background-color: $color-error;
      }
    }
  }
}

.responsive {
  @include respond-to('sm') {
    width: 50px;
    position: absolute;
    right: 0;

    .root {
      .item {
        @include responsive-item;
      }

      .danger {
        @include responsive-item;
      }
    }
  }
}
