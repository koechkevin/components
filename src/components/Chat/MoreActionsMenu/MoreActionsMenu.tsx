import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Menu } from 'antd';
import React, { FC } from 'react';
import { useWindowSize } from 'react-use';

import { MessageItemProps } from '../../../typings/Message';

import styles from './MoreActionsMenu.module.scss';

interface MenuItem {
  icon: any;
  text: string;
  isDanger?: boolean;
  handler: (message: MessageItemProps) => void;
}

interface Props {
  menu: MenuItem[];
  message: MessageItemProps;
  isResponsive?: boolean;
}

const MoreActionsMenu: FC<Props> = (props) => {
  const { isResponsive, message, menu } = props;
  const { width } = useWindowSize();

  if (!menu || menu.length === 0) {
    return null;
  }

  return (
    <div className={isResponsive ? styles.responsive : ''}>
      <Menu className={styles.root}>
        {menu.map(({ icon, text, isDanger, handler }: MenuItem, idx: number) => (
          <Menu.Item
            key={idx}
            className={isDanger ? styles.danger : styles.item}
            onClick={() => message && message.id && handler(message)}
          >
            {width <= 768 ? <FontAwesomeIcon icon={icon} /> : <span>{text}</span>}
          </Menu.Item>
        ))}
      </Menu>
    </div>
  );
};

export default MoreActionsMenu;
