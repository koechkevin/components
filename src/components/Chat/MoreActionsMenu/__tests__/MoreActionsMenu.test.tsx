import { faPencil, faTrash } from '@fortawesome/pro-regular-svg-icons';
import { Menu } from 'antd';
import { mount } from 'enzyme';
import React from 'react';

import MoreActionsMenu from '../MoreActionsMenu';

describe('Tests for MoreActionsMenu component', () => {
  const props = {
    message: {
      id: 'sampleMessageId',
      threadId: 'sampleThreadId',
      createdAt: '2019-10-08 06:22:09',
      updatedAt: '2020-10-08 06:22:09',
      author: {
        avatarColor: '#4224',
        name: 'Jane Doe',
        productId: 'company-portal',
        chatStatus: 'sample chat status',
      },
    },
    menu: [
      {
        icon: faPencil,
        text: 'Edit Message',
        handler: jest.fn(),
      },
      {
        icon: faPencil,
        text: 'Delete Message',
        handler: jest.fn(),
      },
    ],
  };

  const wrapper = mount(<MoreActionsMenu {...props} />);

  it('should render MoreActionsMenu without crushing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should simulate click of menu items', () => {
    const menuItems = wrapper.find(Menu.Item);

    menuItems.forEach((menuItem) => {
      menuItem.simulate('click');
    });

    expect(props.menu[0].handler).toHaveBeenCalled();
    expect(props.menu[1].handler).toHaveBeenCalled();
  });
});
