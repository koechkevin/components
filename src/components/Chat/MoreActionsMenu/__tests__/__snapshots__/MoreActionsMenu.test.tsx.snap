// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for MoreActionsMenu component should render MoreActionsMenu without crushing 1`] = `ReactWrapper {}`;
